core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -286,6 +286,44 @@ private[spark] class TaskSchedulerImpl(
     }
   }
 
+  /**
+   * SPARK-25250: Whenever any Task gets successfully completed, we simply mark the
+   * corresponding partition id as completed in all attempts for that particular stage and
+   * additionally, for a Result Stage, we also kill the remaining task attempts running on the
+   * same partition. As a result, we do not see any Killed tasks due to
+   * TaskCommitDenied Exceptions showing up in the UI. When this method is called from
+   * DAGScheduler.scala on a task completion event being fired, it is assumed that the new
+   * TaskSet has already been created and registered. However, a small possibility does exist
+   * that when this method gets called, possibly the new TaskSet might have not been added
+   * to taskSetsByStageIdAndAttempt. In such a case, we might still hit the same issue. However,
+   * the above scenario has not yet been reproduced.
+   */
+  override def completeTasks(partitionId: Int, stageId: Int, killTasks: Boolean): Unit = {
+    taskSetsByStageIdAndAttempt.getOrElse(stageId, Map()).values.foreach { tsm =>
+      tsm.partitionToIndex.get(partitionId) match {
+        case Some(index) =>
+          tsm.markPartitionAsAlreadyCompleted(index)
```

Comment:
Having another look at `completeTasks` & `markPartitionAsAlreadyCompleted`, given my thoughts below.

First of all, I think we have reached an agreement on killing the running attempts for the same partition.

Besides,

* if `tsm` is active and task corresponding to `partitionId` here is its last unfinished task
      - the task is not running 
        So, only  mark `successful(index) = true` but do not mark it as `zombie` would make it never have a chance to 
        increase `tasksSuccessful` and finish itself later.
      - the task is running
        `tsm` will hanle the killed task at the end, but also can not finish itself. Because we won't mark the TaskSet as 
         `zombie` (unless it's a barrier TaskSet) when we handle failed task. Also, `tasksSuccessful`.

* if `tsm` is zombie
Mark `successful(index) = true` is meaningless. Because we've already sent killing siganls to thoes tasks, so I believe those zombie TaskSetManager would handle them normally at the end.  So, even call `maybeFinishTaskSet()` here is useless.

Furthermore, I'm thinking about whether we need to exclude the `tsm` who succeed the task firstly and trigger all the stuff subsequently, so that we may avoid doing some redundant stuff, e.g. `successful(index) = true`.  **But, we do also need to kill tasks for it.**

cc @cloud-fan @pgandhi999  WDYT ?

Reply:
I don't see any downside of marking `successful(index) = true`.
- If `tsm` is active and the task is pending to run, I think killing the task is a no-op, but maybe better to not send the KillTask request to avoid overhead. `successful(index) = true` is needed here to finish the `tsm`.
- if `tsm` is active and the task is running. Killing the task will probably save resource, and `successful(index) = true` is needed here to finish the `tsm`.
- if `tsm` is zoombie, `successful(index) = true` is a no-op(but no overhead) as a zombie `tsm` will never be marked as finished. Killing the task may have benefit if the task is still running.

If we want to be 100% safe and eliminate any perf regression, maybe the simplest choice is to not kill the task, but just `successful(index) = true`.

Reply:
Hmmm... I'm not saying we can't mark `successful(index) = true`, but we should not only mark  `successful(index) = true` but also mark `isZombie=true` for some cases. Because, a `tsm` can not be finished if it is not zombie.

And to be clarify, again, the key point we're arguing is none business of whether to kill the tasks. We'll kill the tasks after calling `markPartitionAsAlreadyCompleted()`, I promise.

> If `tsm` is active and the task is pending to run, I think killing the task is a no-op, but maybe better to not send the KillTask request to avoid overhead. `successful(index) = true` is needed here to finish the `tsm`.

We should not only mark `successful(index) = true`, but also increase `tasksSuccessful`(this may has a little affect on blacklist) and even `mark the `tsm` as zombie if the task is the last successful task for the `tsm`. otherwise, it can not be finished .

>if `tsm` is active and the task is running. Killing the task will probably save resource, and `successful(index) = true` is needed here to finish the `tsm`.

Same as above.

> if `tsm` is zoombie, `successful(index) = true` is a no-op(but no overhead) as a zombie `tsm` will never be marked as finished. Killing the task may have benefit if the task is still running.

As I mentioned previously, mark `successful(index) = true` is unnecessary. So, I'm fine whether we do this or not.

Correct me if I'm wrong, thanks, @cloud-fan .

Reply:
The framework guarantees that, for a certain stage, there will be one active/finished `tsm` and 0 or more zombie `tsm`. I don't want to complicate it here by setting the zombie bit.

But I get your point, I think we need to increase `tasksSuccessful`, so that a `tsm` can correctly be finished.

Reply:
I know, for  a certain stage, there will be one active `tsm` and 0 or more zombie tsm. But they can all be finished by calling `maybeFinishTaskSet()`, and this only notify `TaskScheduler`, but remains unkonw for `DAGScheduler` (I guess you're meaning this.)

>  I think we need to increase tasksSuccessful, so that a tsm can correctly be finished.

Only increase `tasksSuccessful` is not enough. Because a `tsm` could be finished if and only if it is zombie. And if we don't set the zombie bit here, then, no else where can do this for certain cases I mentioned above. So, the `tsm` won't be finished at the end.

Reply:
I checked the code, you are right!

I think we should extract out the logic of finishing a task from `handleSuccessfulTask`, and call that logic in the new `markPartitionAsAlreadyCompleted`

Reply:
That sounds good.

Reply:
I see your point here. In that case, my opinion is that it will make more sense that I call the existing method `markPartitionCompleted` which already handles the logic of finishing a task instead of creating a new method `markPartitionAsAlreadyCompleted`. Does that sound good @cloud-fan  and @Ngone51 ?

Reply:
Yeah, I think `markPartitionCompleted()` has already extracted out some logic from `handleSuccessfulTask()`. And we can strengthen it if we need more.

Reply:
Ok a couple of pointers here. I was testing the code when calling `markPartitionCompleted()` and noticed that increasing the counter `tasksSuccessful` for all tasks was causing issues as the counter was also increasing for completed tasks. So my thoughts are that:

- We should call `markPartitionCompleted()` only for running tasks and call `markPartitionAsAlreadyCompleted()` for non-running tasks.
- For running tasks, we should call `markPartitionCompleted()` only when we successfully kill them. If the scheduler backend does not support killing tasks, we call `markPartitionAsAlreadyCompleted()`. WDYT @Ngone51 @cloud-fan ?

Reply:
Also, also, while trying the above code, I see the same race condition again, for reasons I am trying to still debug but was wondering whether or not it is a good idea to call `markPartitionCompleted()`.

Reply:
@Ngone51 @cloud-fan I am using `markPartitionCompleted()` now and it works. I have tested it successfully. Thank you.